R/insert-ui.R

```diff
@@ -172,3 +172,275 @@ removeUI <- function(selector,
   if (!immediate) session$onFlushed(callback, once = TRUE)
   else callback()
 }
+
+
+#' Dynamically add a tabPanel
+#'
+#' Dynamically add a \code{tabPanel()} into an existing \code{tabsetPanel}.
+#'
+#' @param tabsetPanelId The \code{id} of the \code{tabsetPanel()} into which
+#'   \code{tab} will be inserted.
+#'
+#' @param tab The tab element to be added (must be created with
+#'   \code{tabPanel}).
+#'
+#' @param target The \code{value} of an existing \code{tabPanel()}, next to
+#'   which \code{tab} will be added. If \code{NULL},the \code{tab} will be
+#'   placed either as the first tab or the last tab, depending on the
+#'   \code{position} argument.
+#'
+#' @param position Should \code{tab} be added to the right ot to the left
+#'   of \code{target}?
+#'
+#' @param immediate Whether \code{tab} should be immediately inserted into
+#'   the app when you call \code{insertTab}, or whether Shiny should wait until
+#'   all outputs have been updated and all observers have been run (default).
+#'
+#' @param session The shiny session within which to call \code{insertTab}.
+#'
+#' @seealso \code{\link{removeTab}}, \code{\link{showTab}},
+#'   \code{\link{hideTab}}
+#'
+#' @examples
+#' ## Only run this example in interactive R sessions
+#' if (interactive()) {
+#' ui <- fluidPage(
+#'   sidebarLayout(
+#'     sidebarPanel(actionButton("add", "Add tab")),
+#'     mainPanel(
+#'       tabsetPanel(id = "tabs",
+#'         tabPanel("Hello", "This is the hello tab"),
+#'         tabPanel("Foo", "This is the foo tab"),
+#'         tabPanel("Bar", "This is the bar tab")
+#'       )
+#'     )
+#'   )
+#' )
+#' server <- function(input, output, session) {
+#'   observeEvent(input$add, {
+#'     insertTab(tabsetPanelId = "tabs",
+#'       tabPanel("Dynamic", "This a dynamically-added tab"),
+#'       target = "Bar"
+#'     )
+#'   })
+#' }
+#'
+#' shinyApp(ui, server)
+#' }
+#' @export
+insertTab <- function(tabsetPanelId, tab, target = NULL,
+  position = c("right", "left"), immediate = FALSE,
+  session = getDefaultReactiveDomain()) {
+
+  force(tabsetPanelId)
+  force(tab)
+  force(target)
+  position <- match.arg(position)
+  force(session)
+
+  iconClass <- tab$attribs$`data-icon-class`
+  icon <- if (!is.null(iconClass)) {
+    # for font-awesome we specify fixed-width
```

Comment:
This was copied from here: (R/bootstrap.R#L818-L823)[https://github.com/rstudio/shiny/blob/master/R/bootstrap.R#L818-L823]. I admit i didn't really look into it, but I can if neither you or @jcheng5 remember this. Let me know!